User crud

aliu19 opened the pull request.

aliu19 committed:
added user role to schema

aliu19 committed:
update user info while keeping session

caijiah commented on controllers/UserController.ts:
```diff
@@ -107,9 +109,35 @@ export default class UserController implements UserControllerI {
+        let existingUser = false
+
+        // checks if username is taken
+        if (userInfo.username) {
```
If a user doesn't update their username, this checking could always fail.
Perhaps we should compare the username from current session.

caijiah replied:
Since I am send all profile to the backend. Thanks

caijiah commented on mongoose/users/UserSchema.ts:
```diff
@@ -36,7 +37,8 @@ const UserSchema = new mongoose.Schema<User>({
     joined: {type: Date, default: Date.now},
-    location: LocationSchema
+    location: LocationSchema,
+    role: {type: String, default: Role.General, enum: Role}
```
Since we have role, could you please remove accountType?

caijiah commented on controllers/UserController.ts:
```diff
@@ -107,9 +109,35 @@ export default class UserController implements UserControllerI {
+            }
+
+            // update the user
+            await UserController.userDao.updateUser(req.params.uid, userInfo).then(status => res.json(status))
```
I will pass the profile with "_id", perhaps we don't need uid to be in the api url for security.

aliu19 replied:
keeping as is for now

aliu19 committed:
removed accountType

aliu19 committed:
made user fields public and updated signup to prof's imp

aliu19 committed:
update login to prof's imp

aliu19 committed:
update user for all fields

since front end passes all fields over to update even though they may not be changed

aliu19 committed:
hide password in session while updating user

aliu19 commented:
Made necessary changes, please check if the session is the same when the username and/or password changes.

caijiah committed:
Fixed the issue that a user didn't update the password

caijiah committed:
Current fix for session data is not updated when calling updateUser. We make sure profile will fetch the latest user's profile data in profile API

averywpx commented on models/users/User.ts:
```diff
@@ -4,6 +4,7 @@
 import AccountType from "./AccountType";
```
You may want to remove it and also the AccountType.ts in user model to avoid confusing. It is not a big deal so I will just approve the merge.

aliu19 committed:
removed accountType import

aliu19 commented:
Jiahao's changes work + okay with front end. Merging.
